File: mscp-to-fleet-yaml/fix_queries.go
package main

import (
	"fmt"
	"os"
	"path/filepath"
	"regexp"
	"strings"
)

// QueryFixer handles fixing generic queries in YAML files
type QueryFixer struct{}

// NewQueryFixer creates a new query fixer
func NewQueryFixer() *QueryFixer {
	return &QueryFixer{}
}

// FixGenericQueries fixes generic queries in a YAML file
func (qf *QueryFixer) FixGenericQueries(filePath string) (int, error) {
	fmt.Printf("Processing %s...\n", filePath)

	content, err := os.ReadFile(filePath)
	if err != nil {
		return 0, fmt.Errorf("failed to read file %s: %w", filePath, err)
	}

	originalContent := string(content)
	changes := 0

	// Fix generic SELECT 1; queries
	// These should be replaced with meaningful queries based on policy context
	select1Pattern := regexp.MustCompile(`(\s+query: SELECT 1;)(\s+purpose: Informational)`)
	originalContent = select1Pattern.ReplaceAllStringFunc(originalContent, func(match string) string {
		changes++
		return strings.Replace(match, "SELECT 1;", "SELECT 1;  # TODO: Replace with specific query for this policy", 1)
	})

	// Fix overly generic file path queries
	filePathPattern := regexp.MustCompile(`(\s+query: SELECT 1 FROM file WHERE path LIKE '/[^%]+%';)(\s+purpose: Informational)`)
	originalContent = filePathPattern.ReplaceAllStringFunc(originalContent, func(match string) string {
		changes++
		return strings.Replace(match, "SELECT 1 FROM file WHERE path LIKE '/[^%]+%';", "SELECT 1 FROM file WHERE path LIKE '/[^%]+%';  # TODO: Replace with specific file validation query", 1)
	})

	// Fix generic launchd queries
	launchdPattern := regexp.MustCompile(`(\s+query: SELECT 1 FROM launchd WHERE name LIKE '%[^%]+%';)(\s+purpose: Informational)`)
	originalContent = launchdPattern.ReplaceAllStringFunc(originalContent, func(match string) string {
		changes++
		return strings.Replace(match, "SELECT 1 FROM launchd WHERE name LIKE '%[^%]+%';", "SELECT 1 FROM launchd WHERE name LIKE '%[^%]+%';  # TODO: Replace with specific service validation query", 1)
	})

	if changes > 0 {
		err = os.WriteFile(filePath, []byte(originalContent), 0644)
		if err != nil {
			return 0, fmt.Errorf("failed to write file %s: %w", filePath, err)
		}
		fmt.Printf("  Added %d TODO comments for generic queries\n", changes)
	} else {
		fmt.Printf("  No generic queries found to fix\n")
	}

	return changes, nil
}

// ProcessAllFiles processes all YAML files in the current directory
func (qf *QueryFixer) ProcessAllFiles() error {
	yamlFiles, err := filepath.Glob("*-fleet-policies.yml")
	if err != nil {
		return fmt.Errorf("failed to find YAML files: %w", err)
	}

	totalChanges := 0

	for _, yamlFile := range yamlFiles {
		if strings.HasSuffix(yamlFile, ".bak") {
			continue
		}

		changes, err := qf.FixGenericQueries(yamlFile)
		if err != nil {
			fmt.Printf("Error processing %s: %v\n", yamlFile, err)
			continue
		}
		totalChanges += changes
	}

	fmt.Printf("\nTotal changes made: %d\n", totalChanges)
	fmt.Println("\nNext steps:")
	fmt.Println("1. Review the TODO comments in each file")
	fmt.Println("2. Replace TODO comments with specific queries based on policy requirements")
	fmt.Println("3. Test the corrected queries")
	return nil
}

// RunFixQueries runs the query fixer
func RunFixQueries() error {
	fixer := NewQueryFixer()
	return fixer.ProcessAllFiles()
}

File: mscp-to-fleet-yaml/fix_queries.py
#!/usr/bin/env python3
"""
Script to fix generic queries in Fleet policy YAML files.
This script identifies and corrects common problematic query patterns.
"""

import os
import re
import sys
from pathlib import Path

def fix_generic_queries(file_path):
    """Fix generic queries in a YAML file."""
    print(f"Processing {file_path}...")
    
    with open(file_path, 'r') as f:
        content = f.read()
    
    original_content = content
    
    # Fix generic SELECT 1; queries
    # These should be replaced with meaningful queries based on policy context
    content = re.sub(
        r'(\s+query: SELECT 1;)(\s+purpose: Informational)',
        r'\1  # TODO: Replace with specific query for this policy\2',
        content
    )
    
    # Fix overly generic file path queries
    content = re.sub(
        r'(\s+query: SELECT 1 FROM file WHERE path LIKE \'/[^%]+%\';)(\s+purpose: Informational)',
        r'\1  # TODO: Replace with specific file validation query\2',
        content
    )
    
    # Fix generic launchd queries
    content = re.sub(
        r'(\s+query: SELECT 1 FROM launchd WHERE name LIKE \'%[^%]+%\';)(\s+purpose: Informational)',
        r'\1  # TODO: Replace with specific service validation query\2',
        content
    )
    
    # Count changes made
    changes = len(re.findall(r'# TODO:', content)) - len(re.findall(r'# TODO:', original_content))
    
    if changes > 0:
        with open(file_path, 'w') as f:
            f.write(content)
        print(f"  Added {changes} TODO comments for generic queries")
        return changes
    else:
        print(f"  No generic queries found to fix")
        return 0

def main():
    """Main function to process all YAML files."""
    yaml_dir = Path('.')
    yaml_files = list(yaml_dir.glob('*-fleet-policies.yml'))
    
    total_changes = 0
    
    for yaml_file in yaml_files:
        if yaml_file.name.endswith('.bak'):
            continue
            
        changes = fix_generic_queries(yaml_file)
        total_changes += changes
    
    print(f"\nTotal changes made: {total_changes}")
    print("\nNext steps:")
    print("1. Review the TODO comments in each file")
    print("2. Replace TODO comments with specific queries based on policy requirements")
    print("3. Test the corrected queries")

if __name__ == '__main__':
    main()
